### PR TITLE
TDL-19174 skipped the record for out of range date values

### PR DESCRIPTION
**Pull request by namrata270998**
# Description of change
- Skipped the record, when SchemaMismatch is thrown due to date out of range value error

# Manual QA steps
 - Verify that the record is skipped for out of range date values
 - Verify that the record is successfully written for valid date values
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_jira/streams.py [hunk 1]
```diff
@@ -4,11 +4,48 @@
...
 DEFAULT_PAGE_SIZE = 50
 
+def handle_date_time_schema_miss_match(exception, record): # pylint: disable=inconsistent-return-statements
+    """
+    Handling exception for date-time value out of range.
+    """
+
+    if ("{'format': 'date-time', 'type': ['string', 'null']}"
+        in exception.args[0].split("\n\t")[1]):
+        nested_keys = (
```
Comment (dbshah1212) [inline tap_jira/streams.py:21]:
```suggestion
        nested_keys = exception.args[0].split("\n\t")[1].split(":")[0]
```

File: tap_jira/streams.py [hunk 2]
```diff
@@ -4,11 +4,48 @@
...
+            exception.args[0].split("\n\t")[1].split(":")[0]
+        )
+        # Getting the date value from nested object
+        obj = record.copy()
```
Comment (dbshah1212) [inline tap_jira/streams.py:25]: comment mismatch: We are not getting the date value here.

File: tap_jira/streams.py [hunk 3]
```diff
@@ -4,11 +4,48 @@
...
+        for key in nested_keys.split("."):
+            # In nested object to convert list index to integer
```
Comment (dbshah1212) [inline tap_jira/streams.py:27]:
```suggestion
            # Convert list index of nested object to integer
```

File: tap_jira/streams.py [hunk 4]
```diff
@@ -4,11 +4,48 @@
...
+            obj = obj[int(key) if key.isnumeric() else key]
+
+        try:
+            # Parsing date to catch 'out of range' error
+            utils.strptime_to_utc(obj)
+        except ParserError as err:
+
+            # Check the error message if the 'year' or 'day' is out of range
+            if ("out of range" in str(err)) or (
+                # Check the error message if 'month' or ['hours','minutes','seconds'] given in date is not in range
```
Comment (dbshah1212) [inline tap_jira/streams.py:37]: Give a sample error example here.

File: tap_jira/streams.py [hunk 5]
```diff
@@ -4,11 +4,48 @@
...
+                "must be in" in str(err)):
+                LOGGER.warning("Skipping record of id: %s due to Date out of range, DATE: %s", record["id"], obj)
+                return True
+            else:
+                # raise an error if exception is not for 'out of range' date
+                raise err
+    else:
+        # raise an error in case of schema mismatch error other than ones
```
Comment (dbshah1212) [inline tap_jira/streams.py:45]:
```suggestion
        # raise a schema mismatch error, other than date out of range values.
```

File: tap_jira/streams.py [hunk 6]
```diff
@@ -110,7 +147,12 @@ def write_page(self, page):
         extraction_time = singer.utils.now()
         for rec in page:
             with Transformer() as transformer:
-                rec = transformer.transform(rec, stream.schema.to_dict(), stream_metadata)
+                try:
+                    rec = transformer.transform(rec, stream.schema.to_dict(), stream_metadata)
+                except SchemaMismatch as ex:
+                    # Checking if schema-mismatch is occurring for datetime value
```
Comment (dbshah1212) [inline tap_jira/streams.py:153]: Provide a card and detailed description here in a comment.

File: tap_jira/streams.py [hunk 7]
```diff
@@ -4,11 +4,48 @@
...
+        for key in nested_keys.split("."):
+            # In nested object to convert list index to integer
+            obj = obj[int(key) if key.isnumeric() else key]
+
+        try:
+            # Parsing date to catch 'out of range' error
+            utils.strptime_to_utc(obj)
+        except ParserError as err:
+
+            # Check the error message if the 'year' or 'day' is out of range
```
[the remark this answers is not in the document]
Comment (namrata270998) [reply tap_jira/streams.py:34]: Addedan example

File: tap_jira/streams.py [hunk 8]
```diff
@@ -4,11 +4,48 @@
...
+            # example: year 51502 is out of range: 51502-06-08T14:46:42.000000
+            if ("out of range" in str(err)) or (
+                # Check the error message if 'month' or ['hours','minutes','seconds'] given in date is not in range
+                # example: month must be in 1..12: 5150-33-08T14:46:42.000000
+                "must be in" in str(err)):
+                LOGGER.warning("Skipping record of id: %s due to Date out of range, DATE: %s", record["id"], obj)
```
Comment (umangagrawal-crest) [inline tap_jira/streams.py:40]: The code will throw "KEYERROR", for the streams which do not contain id field

File: tap_jira/streams.py (hunk 2 above)
Comment (namrata270998) [reply tap_jira/streams.py:23]: Added proper comment

File: tap_jira/streams.py (hunk 1 above)
Comment (namrata270998) [reply tap_jira/streams.py]: No we dont, removed the brackets

File: tap_jira/streams.py (hunk 4 above)
Comment (namrata270998) [reply tap_jira/streams.py:37]: Added an error example in the comment

File: tap_jira/streams.py (hunk 5 above)
Comment (namrata270998) [reply tap_jira/streams.py]: Updated

File: tap_jira/streams.py (hunk 6 above)
Comment (namrata270998) [reply tap_jira/streams.py:153]: Added the card description with the card link

File: tap_jira/streams.py (hunk 3 above)
Comment (namrata270998) [reply tap_jira/streams.py]: Updated

File: tap_jira/streams.py [hunk 9]
```diff
@@ -4,11 +4,48 @@
...
 from .context import Context
 
 DEFAULT_PAGE_SIZE = 50
 
+def handle_date_time_schema_miss_match(exception, record): # pylint: disable=inconsistent-return-statements
+    """
+    Handling exception for date-time value out of range.
+    """
+
+    if ("{'format': 'date-time', 'type': ['string', 'null']}"
```
Comment (dbshah1212) [inline tap_jira/streams.py:19]:
```suggestion
    if ("{'format': 'date-time', 'type': ['null', 'string']}"
```

Comment (dbshah1212) [inline tap_jira/streams.py:19]: Each time we will get String first then null? or are there any chances where we might get null first then string?

Comment (namrata270998) [reply tap_jira/streams.py:19]: We've verified and observed that we're getting string first and then null for the type in the error message.

File: tap_jira/streams.py (hunk 8 above)
Comment (namrata270998) [reply tap_jira/streams.py]: Updated the code and used the id dynamically instead of static values.

File: tap_jira/streams.py (hunk 9 above)
Comment (dbshah1212) [reply tap_jira/streams.py:19]: For each stream?

Comment (dbshah1212) [reply tap_jira/streams.py:19]: If we have a different structure then we must modify that in the catalog.

File: tap_jira/streams.py [hunk 10]
```diff
@@ -4,11 +4,48 @@
...
+            # Check the error message if the 'year' or 'day' is out of range
+            # example: year 51502 is out of range: 51502-06-08T14:46:42.000000
+            if ("out of range" in str(err)) or (
+                # Check the error message if 'month' or ['hours','minutes','seconds'] given in date is not in range
+                # example: month must be in 1..12: 5150-33-08T14:46:42.000000
+                "must be in" in str(err)):
+                LOGGER.warning("Skipping record of: %s due to Date out of range, DATE: %s", dict((pk, record.get(pk)) for pk in pk_fields), obj)
+                return True
+            else:
+                # raise an error if exception is not for 'out of range' date
```
Comment (dbshah1212) [inline tap_jira/streams.py:43]:
```suggestion
                # Raise an error for exception except 'out of range' date
```

File: tests/unittests/test_date_out_of_range.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    @mock.patch("tap_jira.streams.singer.utils.now", return_value="2022-05-23T09:16:11.356670Z")
+    @mock.patch("tap_jira.streams.singer.write_record")
+    @mock.patch("tap_jira.streams.Context")
+    @mock.patch("tap_jira.streams.metadata")
+    def test_out_of_range_date(self, mock_metadata, mock_Context, mock_write_record, mock_now):
+        """
+        Verify that the tap successfully skips record for out of range date value.
+        """
+        mock_records =[
+            {"updated": "2017000-09-05T19:51:03.159Z", "id": "1","id2": "11"},  # year out of range
```
Comment (dbshah1212) [inline tests/unittests/test_date_out_of_range.py:15]: Have we take any sample of `month must be in 1..12`?

File: tap_jira/streams.py (hunk 9 above)
Comment (namrata270998) [reply tap_jira/streams.py:19]: Yes, we are getting the same error for each stream.

File: tests/unittests/test_date_out_of_range.py (hunk 11 above)
Comment (namrata270998) [reply tests/unittests/test_date_out_of_range.py:15]: Yes, verified the same. Please see [L17](https://github.com/singer-io/tap-jira/pull/87/files#diff-38c5ac42c3c3fcae142935dae6ed190e2647fea82e6808ebc0ced96055a3387bR17) for reference

File: tap_jira/streams.py (hunk 10 above)
Comment (namrata270998) [reply tap_jira/streams.py]: Updated the comment

File: tap_jira/streams.py (hunk 9 above)
Comment (dbshah1212) [reply tap_jira/streams.py:19]: ok

**Review by dbshah1212 — APPROVED** [review]

**Review by umangagrawal-crest — APPROVED** [review]

File: tap_jira/streams.py [hunk 12]
```diff
@@ -4,11 +4,48 @@
...
 
 from singer import metrics, utils, metadata, Transformer
+from singer.transform import SchemaMismatch
+from dateutil.parser._parser import ParserError
 from .http import Paginator,JiraNotFoundError
 from .context import Context
 
 DEFAULT_PAGE_SIZE = 50
 
+def handle_date_time_schema_miss_match(exception, record, pk_fields): # pylint: disable=inconsistent-return-statements
```
Comment (savan-chovatiya) [inline tap_jira/streams.py:14]:
```suggestion
def handle_date_time_schema_mis_match(exception, record, pk_fields): # pylint: disable=inconsistent-return-statements
```

Comment (namrata270998) [reply tap_jira/streams.py]: Fixed the typo

**Review by prijendev — APPROVED** [review]

**Review by savan-chovatiya — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]

File: tests/unittests/test_date_out_of_range.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            {"updated": "2001-09-05T19:51:03.159000", "id": "2","id2": "12"},  # Correct record
+            {"updated": "2001-13-05T19:51:03.159000", "id": "3","id2": "13"},  # month out of range
+            {"updated": "2007-09-45T19:51:03.159000z", "id": "4","id2": "14"},  # day out of range
+            {"updated": "2006-01-29T01:62:01.99999", "id": "5","id2": "16"},  # minutes out of range
+            {"updated": "2011-09-05T19:51:03.159000", "id": "6","id2": "15"},  # Correct record
+            {"updated": "2009-09-10T26:51:03.159000Z", "id": "7","id2": "17"},  # hour out of range
+        ]
```
Comment (RushT007) [inline tests/unittests/test_date_out_of_range.py:22]: I think last record should be valid one to verify that all records skipped with invalid format and all valid ones replicated.

**Review by RushT007 — CHANGES_REQUESTED** [review]

File: tests/unittests/test_date_out_of_range.py (hunk 13 above)
Comment (namrata270998) [reply tests/unittests/test_date_out_of_range.py:22]: Yes, thats right, hence we have just 2 call counts [here](https://github.com/singer-io/tap-jira/pull/87/files#diff-38c5ac42c3c3fcae142935dae6ed190e2647fea82e6808ebc0ced96055a3387bR39)

Comment (namrata270998) [reply tests/unittests/test_date_out_of_range.py]: Updated the unittest as per your suggestions

**Review by shantanu73 — APPROVED** [review]

**Review by dsprayberry — APPROVED** [review]

**Pull request merged**